query/src/main/java/org/infinispan/query/affinity/AffinityIndexManager.java

```diff
@@ -155,26 +179,23 @@ private void handleOwnershipAcquired() {
 
    @Override
    public void flushAndReleaseResources() {
-      Address lockHolder = getLockHolder(getIndexName());
-      if (lockHolder != null) {
-         InfinispanDirectoryProvider directoryProvider = (InfinispanDirectoryProvider) getDirectoryProvider();
-         int activeDeleteTasks = directoryProvider.getActiveDeleteTasks();
-         boolean wasInterrupted = false;
-         while (activeDeleteTasks > 0) {
-            try {
-               Thread.sleep(POLL_WAIT);
-               logDebug("Waiting for pending delete tasks, remaining: %s", activeDeleteTasks);
-               activeDeleteTasks = directoryProvider.getActiveDeleteTasks();
-            } catch (InterruptedException e) {
-               wasInterrupted = true;
-            }
-         }
-         if (wasInterrupted) {
-            Thread.currentThread().interrupt();
+      InfinispanDirectoryProvider directoryProvider = (InfinispanDirectoryProvider) getDirectoryProvider();
+      int activeDeleteTasks = directoryProvider.pendingDeleteTasks();
+      boolean wasInterrupted = false;
+      while (activeDeleteTasks > 0) {
+         try {
+            Thread.sleep(POLL_WAIT);
```

Comment:
@slaskawi @gustavonalle In theory you could have a "deadlock" if a remote node throws an `OutdatedTopologyException` during the remove, but the local node can't install the newer topology because you're blocking it. These removals are triggered from a `@TopologyChanged` listener, which means they do block new topology updates (`LocalTopologyManagerImpl` uses a `SemaphoreCompletionService` to serialize topology updates, including the notifications). OTOH, `StateTransferInterceptor` doesn't block forever, it times out after `clustering.remoteTimeout`, so it's not a real deadlock.